src/brevitas/core/stats/stats_op.py

```diff
@@ -530,52 +587,29 @@ def evaluate_loss(self, x, candidate):
         return loss
 
-    def mse_grid_search(self, xl, x):
-        best_loss = torch.tensor(float('inf'), device=x.device, dtype=x.dtype)
-        best_candidate = xl
-        for i in range(2, self.num + 1):
-            candidate = (xl * i).detach()
-            loss = self.evaluate_loss(x, candidate)
-            best_candidate = torch.where(loss < best_loss, candidate, best_candidate)
-            best_loss = torch.min(loss, best_loss)
-        return best_candidate
-
-    def mse_fib_search(self, xl, xr, x):
-
-        def fib_seq(n):
-            if n <= 0:
-                return [0]
-            seq = [0, 1]
-            while len(seq) <= n:
-                next = seq[-1] + seq[-2]
-                seq.append(next)
-            return seq
+    def _mse_search(self, xl, xr, loss_fn, num_iter):
+        if self.search_method == 'grid':
+            return mse_grid_search(xl=xl, xr=xr, loss_fn=loss_fn, num_iter=num_iter)
+        elif self.search_method == 'fibonacci':
+            return mse_fib_search(xl=xl, xr=xr, loss_fn=loss_fn, num_iter=num_iter)
 
-        # vectorized variant of
-        # https://indrag49.github.io/Numerical-Optimization/solving-one-dimensional-optimization-problems.html#fibonacci-search-method
-        F = fib_seq(self.num)
-        L0 = xr - xl
-        Li = (F[self.num - 2] / F[self.num]) * L0
-        for i in range(2, self.num + 1):
-            x1 = torch.where(Li > L0 / 2, xr - Li, xl + Li)
-            x2 = torch.where(Li > L0 / 2, xl + Li, xr - Li)
-            f1, f2 = self.evaluate_loss(x, x1), self.evaluate_loss(x, x2)
-            xr = torch.where(f1 <= f2, x2, xr)
-            xl = torch.where(f1 >= f2, x1, xl)
-            Li = (F[self.num - i] / F[self.num - (i - 2)]) * torch.where(f1 != f2, L0, xr - xl)
-            L0 = xr - xl
-        return torch.where(f1 <= f2, x1, x2)
+        raise ValueError(f"Search method {self.search_method} not supported.")
 
     def mse_search(self, x):
         x_view = self.input_view_shape_impl(x)
-        init = self.mse_init_op(x_view).detach()
+        init = torch.abs(self.mse_init_op(x_view)).detach()
         base = init / self.num
-        if self.search_method == 'grid':
-            best_candidate = self.mse_grid_search(base, x)
-        elif self.search_method == 'fibonacci':
-            best_candidate = self.mse_fib_search(base, init, x)
-        else:
-            raise ValueError(f"Search method {self.search_method} not supported.")
+        loss_fn = lambda candidate: self.evaluate_loss(x, candidate)
+        best_candidate, best_loss = self._mse_search(xl=base, xr=init, loss_fn=loss_fn, num_iter=self.num if self.is_scale_unsigned else self.num // 2)
+        if not self.is_scale_unsigned:
+            best_neg_candidate, best_neg_loss = self._mse_search(
+                xl=-init,
+                xr=-base,
+                loss_fn=loss_fn,
+                num_iter=self.num // 2,
+            )
+            best_candidate = torch.where(
```

Comment:
In the case of `fib_search`, having it split this way is not equivalent to a single search (as if the values are all positive), or am I missing something?